readme.md
# ECDSA Key Generator - [ecdsa_keygen.py](ecdsa_keygen.py)

This Python script generates Elliptic Curve Digital Signature Algorithm (ECDSA) key pairs. It supports generating keys of varying strengths and provides multiple options for output.

Author: Sean Rice, ecdsagen@programmerscache.com

Current Version: 0.1.1

## Features

- Generation of ECDSA key pairs.
- Supports multiple elliptic curves: SECP256R1, SECP384R1, and SECP521R1.
- Ability to use multiprocessing for faster key generation.
- Options for outputting keys in JSON format or saving to files.
- Tracking of average key generation time.
- Optional password encryption for key files.

## Requirements

Before running the script, ensure you have the required packages installed:

```
pip install -r requirements.txt
```
- Requires Python 3.6 or higher.
- Requires the [cryptography](https://pypi.org/project/cryptography/) package to use SECP256R1, SECP384R1, and SECP521R1 elliptic curves for key generation.

## Usage

Run the script from the command line with the following arguments:

```
python ecdsa_keygen.py [OPTIONS]
```

## Options

- `--processes`: Number of processes to use (default is the number of CPU cores).
- `--cpu`: Use CPU for key generation (default option).
- <span style="color: red;">PLANNED</span>  `--cuda`: Use CUDA for key generation.
- <span style="color: red;">PLANNED</span> `--opencl`: Use OpenCL for key generation.
- `--file-out`: Specify the output file name for the keys.
- `--encrypt`: Provide an encryption passcode for the key files.
- `--out-count`: Set the number of key pairs to generate (default is 1).
- `--strength`: Select the strength level (0: SECP256R1, 1: SECP384R1, 2: SECP521R1).
- `--json`: Output keys in JSON format.
- `--continuous`: Continuously generate key pairs until the script is stopped unless paired with `--out-count`.

## Global Options
`OUTPUT_FOLDER`: - The folder where the key files will be saved. Defaults to the `output` directory.

`CONSOLE_KEYPAIR_DELIMITER` - The delimiter used to separate the private and public keys when outputting to the console. Defaults to:

`
"----------------------------------------------------------------"
`

(The standard width of the keypairs in the console).

`DISPLAY_CONSOLE_DELIMITER` - Whether or not to display the delimiter when outputting to the console. Defaults to `True`.

`TRACK_TIMING` - Whether or not to track the average key generation time. Defaults to `True`.

## Examples

Generate a single key pair using SECP256R1 and output to the console:

```
python ecdsa_keygen.py
```
Example Output:

```
-----BEGIN PRIVATE KEY-----
MIGHAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBG0wawIBAQQgH4BjBCP01ZaWnYhj
j2fzzOZoOqupoBfopyVwI1B9YNahRANCAARabKrrXm1Y5DeEKJOkdRxcBHTaEBAE
LpbAJH++cWSfEe5YP98z/2TMJ1NYRJlkXPll7y1acoNi9WNNmpiSOquY
-----END PRIVATE KEY-----

-----BEGIN PUBLIC KEY-----
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEWmyq615tWOQ3hCiTpHUcXAR02hAQ
BC6WwCR/vnFknxHuWD/fM/9kzCdTWESZZFz5Ze8tWnKDYvVjTZqYkjqrmA==
-----END PUBLIC KEY-----
```
*Note, a delimiter will not be displayed when only a single key pair is generated, or when `DISPLAY_CONSOLE_DELIMITER` is set to `False`.*

Generate 5 key pairs using SECP384R1 and save to files with encryption:

```
python ecdsa_keygen.py --strength 1 --out-count 5 --file-out mykey --encrypt mypassword
```
Generate 10 key pairs using SECP521R1 and output in JSON format:

```
python ecdsa_keygen.py --strength 2 --out-count 10 --json
```
Example output: [keys.json](output/keys.json)

`--file-out` and `--json` can be used together to save the keys to files and output the keys in JSON format with a custom filename.
```
python ecdsa_keygen.py --strength 2 --out-count 10 --file-out keytest --json
```
Example output: [keytest.json](output/keytest.json)


## Output

- Keys are output to the console by default.
- When `--file-out` is used, keys are saved to separate `.pem` files in the `output` directory.
- If `--json` is selected, a JSON file containing the keys is created in the `output` directory.

<br>
<br>
<br >

# Deserialize JSON to PEM Key Pair Generator - [deserialize_json.py](deserialize_json.py)
This script, deserialize_json.py, is designed to take a JSON file containing ECDSA key pairs and generate PEM files for each key pair.

## Features

- Loads key pairs from a JSON file.
- Adds appropriate PEM headers to both private and public keys.
- Saves the keys as separate PEM files in a specified directory.

## Usage

Run the script from the command line with the following arguments:

```
python deserialize_json.py --json-file <path_to_json_file> [--encrypt-key <encryption_key>] [--out-dir <output_directory>]
```

## Arguments

- `--json-file`: The path to the JSON file containing the key pairs.
- `--encrypt-key`: (Optional) The encryption key, if the private keys in the JSON file are encrypted.
- `--out-dir`: (Optional) The output directory where the PEM files will be saved. Defaults to the current directory.

## Example

Convert keys from a JSON file and save the PEM files to a specific directory:

```
python deserialize_json.py --json-file keys.json --out-dir ./pem_keys
```

This will read the key pairs from keys.json and save the PEM files in the ./pem_keys directory.

## Output
- Private keys are saved with the filename format [index]-priv.pem.
- Public keys are saved with the filename format [index]-pub.pem.


## License

These scripts are released under the [Mozilla Public License 2.0](https://www.mozilla.org/en-US/MPL/2.0/) - [License File](LICENSE).

## Why?
To answer simply, why not?

While I understand it's essentially reinventing the wheel, I wanted to create a simple script that would allow me to generate ECDSA key pairs for use in my own projects; as well as to learn more about the ECDSA algorithm and how it works.  This is a nice standard way to generate bulk ECDSA key pairs, that can be delivered in a variety of manners.

Enjoy responsibly :)

ChangeLog.md
# v 0.1.0
Initial commit
- Generation of ECDSA key pairs.
- Supports multiple elliptic curves: SECP256R1, SECP384R1, and SECP521R1.
- Ability to use multiprocessing for faster key generation.
- Options for outputting keys in JSON format or saving to files.
- Tracking of average key generation time.
- Optional password encryption for key files.
- CPU support
- Deserialization of keys from JSON

# v 0.1.1
- Fix password encryption
- Continuous key generation
- Fix key generation time tracking
- Parallel key generation

deserialize_json.py
import argparse
import json
import os
import base64
from cryptography.fernet import Fernet
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.backends import default_backend

INPUT_FOLDER = "output"
OUTPUT_FOLDER = "keys_output"

def generate_fernet_key(password, salt):
    # Generates a Fernet key based on the given password and salt
    kdf = PBKDF2HMAC(
        algorithm=hashes.SHA256(),
        length=32,
        salt=salt,
        iterations=100000,
        backend=default_backend()
    )
    key = base64.urlsafe_b64encode(kdf.derive(password.encode()))
    return key

def decrypt_data(encrypted_data, password):
    salt = encrypted_data[:16]
    encrypted_data = encrypted_data[16:]

    key = generate_fernet_key(password, salt)
    fernet = Fernet(key)

    # Decrypt the data
    decrypted_data = fernet.decrypt(encrypted_data).decode()

    return decrypted_data


def load_json_keys(filename, password=None):
    filename = os.path.join(INPUT_FOLDER, filename)
    with open(filename, 'rb') as file:
        data = file.read()
        if password is None:
            return json.loads(data)
        else: # Decrypt the data
            decrypted_data = decrypt_data(data, password)
            try:
                return json.loads(decrypted_data)
            except json.JSONDecodeError as e:
                raise ValueError(f"Decryption successful but data is not valid JSON. {e}") from e

def add_pem_headers(key_str, is_private):
    if is_private:
        return "-----BEGIN PRIVATE KEY-----\n" + key_str + "\n-----END PRIVATE KEY-----"
    else:
        return "-----BEGIN PUBLIC KEY-----\n" + key_str + "\n-----END PUBLIC KEY-----"

def save_key_to_file(key_str, filename, out_dir='.'):
    # Create output directory if it doesn't exist
    if not os.path.exists(out_dir):
        os.makedirs(out_dir)
    filename = os.path.join(out_dir, filename)
    with open(filename, 'w', encoding='utf-8') as file:
        file.write(key_str)

def main():
    parser = argparse.ArgumentParser(description='PEM Key Pair Generator from JSON')
    parser.add_argument('--json-file', type=str, required=True, help='JSON file containing key pairs')
    parser.add_argument('--decrypt', type=str, help='Encryption key, if the private keys are encrypted')
    parser.add_argument('--out-dir', type=str, default='output', help='Output directory')
    args = parser.parse_args()
    
    if args.decrypt:
        key_pairs = load_json_keys(args.json_file, password=args.decrypt)
    else:
        key_pairs = load_json_keys(args.json_file)
    
    OUTPUT_FOLDER = args.out_dir
    if not os.path.exists(OUTPUT_FOLDER):
            os.makedirs(OUTPUT_FOLDER)
        
    for i, (priv_key, pub_key) in enumerate(key_pairs.items(), start=1):
        priv_pem = add_pem_headers(priv_key, is_private=True)
        pub_pem = add_pem_headers(pub_key, is_private=False)
        save_key_to_file(priv_pem, f"{i}-priv.pem", out_dir=OUTPUT_FOLDER)
        save_key_to_file(pub_pem, f"{i}-pub.pem", out_dir=OUTPUT_FOLDER)

if __name__ == "__main__":
    main()


ecdsa_gen.py
import argparse
import os
import json
import time
from multiprocessing import Pool
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.backends import default_backend
from cryptography.fernet import Fernet
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
from cryptography.hazmat.primitives.serialization import Encoding, PrivateFormat, NoEncryption
import base64

#################################### Global Variables ##############################################
OUTPUT_FOLDER = "output"
CONSOLE_KEYPAIR_DELIMITER = "----------------------------------------------------------------"
DISPLAY_CONSOLE_DELIMITER = True
TRACK_TIMING = True
####################################################################################################

def generate_ecdsa_keys_sync(strength):
    if strength == 0:
        curve = ec.SECP256R1()
    elif strength == 1:
        curve = ec.SECP384R1()
    elif strength == 2:
        curve = ec.SECP521R1()
    else:
        raise ValueError("Invalid strength value. Choose 0, 1, or 2.")

    private_key = ec.generate_private_key(curve, default_backend())
    public_key = private_key.public_key()

    # Convert to PEM format
    private_pem = private_key.private_bytes(
        encoding=Encoding.PEM,
        format=PrivateFormat.PKCS8,
        encryption_algorithm=NoEncryption()
    ).decode()
    
    public_pem = public_key.public_bytes(
        encoding=Encoding.PEM,
        format=serialization.PublicFormat.SubjectPublicKeyInfo
    ).decode()

    return private_pem, public_pem

def generate_fernet_key(password, salt):
    kdf = PBKDF2HMAC(
        algorithm=hashes.SHA256(),
        length=32,
        salt=salt,
        iterations=100000,
        backend=default_backend()
    )
    key = base64.urlsafe_b64encode(kdf.derive(password.encode()))
    return key

def encrypt_data(data, password):
    # Ensure data is a JSON string
    salt = os.urandom(16)
    key = generate_fernet_key(password, salt)
    fernet = Fernet(key)

    # Encrypt the data and return the salt and encrypted data
    return salt + fernet.encrypt(data.encode())


def save_keys_to_file_str(filename, private_pem, public_pem, password=None):
    filename = os.path.join(OUTPUT_FOLDER, filename)
    data_to_write = private_pem + public_pem

    if password is not None:
        data_to_write = encrypt_data(data_to_write, password)

    with open(filename, 'wb') as file:
        file.write(data_to_write)


def main():
    parser = argparse.ArgumentParser(description='ECDSA Key Generator')
    parser.add_argument('--processes', type=int, default=os.cpu_count(), help='Number of processes to use')
    parser.add_argument('--cpu', action='store_true', help='Use CPU for key generation')
    parser.add_argument('--cuda', action='store_true', help='Use CUDA for key generation')
    parser.add_argument('--opencl', action='store_true', help='Use OpenCL for key generation')
    parser.add_argument('--file-out', type=str, help='Output file name')
    parser.add_argument('--encrypt', type=str, help='Encryption passcode')
    parser.add_argument('--out-count', type=int, default=1, help='Number of key pairs to generate')
    parser.add_argument('--strength', type=int, default=0, choices=[0, 1, 2], help='Strength level (0: SECP256R1, 1: SECP384R1, 2: SECP521R1)')
    parser.add_argument('--json', action='store_true', help='Output keys in JSON format')
    parser.add_argument('--continuous', action='store_true', help='Continuous output of the key pairs to the console')
    parser.add_argument('--rate-limit', type=float, help='Rate limit for continuous output, accepts float values (e.g. 0.5 for 500ms)')
    args = parser.parse_args()

    if not (args.cuda or args.opencl):
        args.cpu = True
    # Estimate time for completion
    avg_time_per_key = get_average_time(str(args.strength))
    estimated_time = avg_time_per_key * args.out_count
    if avg_time_per_key > 0:
        keys_per_second = 1 / avg_time_per_key
        print(f"Average Keys/Second: {keys_per_second:.2f}")
    else:
        print("Average Keys/Second: Not available (no previous data)")

    print(f"Estimated time to completion: {estimated_time:.2f} seconds")
    
    # Timing the key generation
    if TRACK_TIMING:
        start_time = time.time()
    
    def output_keys(key_pair):
        print(f"{key_pair}".strip("()").replace("'", ""))
        if DISPLAY_CONSOLE_DELIMITER and ( args.out_count > 1 ):
            print(CONSOLE_KEYPAIR_DELIMITER)
    
    if args.continuous:
        print("Continuous output enabled")
        if args.out_count > 1:
            print(f"Output count is greater than 1, outputting {args.out_count} keys to the console")
            with Pool(processes=args.processes) as pool:
                key_count = 0
                while key_count < args.out_count:
                    pool.apply_async(generate_ecdsa_keys_sync, (args.strength,), callback=output_keys)
                    key_count += 1
                    if args.rate_limit:
                        time.sleep(args.rate_limit)
                    if key_count == args.out_count:
                        pool.close()
                        pool.join()
            exit(0)
        else:
            print("Outputting keys in serial to the console until stopped")
            with Pool(processes=args.processes) as pool:
                while True:
                    pool.apply_async(generate_ecdsa_keys_sync, (args.strength,), callback=output_keys)
                    if args.rate_limit:
                        time.sleep(args.rate_limit)
    else:
        with Pool(processes=args.processes) as pool:
            key_pairs = pool.starmap(generate_ecdsa_keys_sync, [(args.strength,) for _ in range(args.out_count)])
        
    if TRACK_TIMING:
        end_time = time.time()
    keys = {}
    for i, (private_pem, public_pem) in enumerate(key_pairs):
        if args.json:
            # strip the header and footer from each key
            private_pem = private_pem.replace('-----BEGIN PRIVATE KEY-----\n', '').replace('-----END PRIVATE KEY-----\n', '').strip()
            public_pem = public_pem.replace('-----BEGIN PUBLIC KEY-----\n', '').replace('-----END PUBLIC KEY-----\n', '').strip()
            keys[private_pem] = public_pem
        elif args.file_out:
            filename = f"{args.file_out}-{i+1}.pem"
            save_keys_to_file_str(filename, private_pem, public_pem, args.encrypt)
        else:
            print(private_pem)
            print(public_pem.replace('-----END PUBLIC KEY-----\n', '-----END PUBLIC KEY-----'))
            if DISPLAY_CONSOLE_DELIMITER and ( args.out_count > 1 ):
                print(CONSOLE_KEYPAIR_DELIMITER)

    if args.json:
        if args.file_out:
            json_filename = os.path.join(OUTPUT_FOLDER, f"{args.file_out}.json")
        else:
            json_filename = os.path.join(OUTPUT_FOLDER, "keys.json")
        
        if args.encrypt:
            encrypted_data = encrypt_data(json.dumps(keys), args.encrypt)
            with open(json_filename, 'wb') as json_file:
                json_file.write(encrypted_data)
                
        else:
            with open(json_filename, 'w', encoding='utf-8') as json_file:
                json.dump(keys, json_file, indent=4)

    # Update timing.json
    if TRACK_TIMING:
        total_time = end_time - start_time
        average_time = total_time / args.out_count
        update_timing_json(str(args.strength), average_time)
    
def update_timing_json(strength, new_time):
    timing_data = {}
    if os.path.exists('timing.json'):
        with open('timing.json', 'r', encoding='utf-8') as file:
            timing_data = json.load(file)

    if strength in timing_data:
        count, avg_time = timing_data[strength]
        new_avg = ((avg_time * count) + new_time) / (count + 1)
        timing_data[strength] = [count + 1, new_avg]
    else:
        timing_data[strength] = [1, new_time]

    with open('timing.json', 'w', encoding='utf-8') as file:
        json.dump(timing_data, file, indent=4)

    return timing_data[strength][1]

def get_average_time(strength):
    if os.path.exists('timing.json'):
        with open('timing.json', 'r', encoding='utf-8') as file:
            timing_data = json.load(file)
        return timing_data.get(str(strength), [0, 0])[1]
    return 0

if __name__ == "__main__":
    if not os.path.exists(OUTPUT_FOLDER):
        os.makedirs(OUTPUT_FOLDER)
    main()
